### ISSUE TITLE
[Question] What is the difference between a custom feature extractor and a custom policy?

### ISSUE THREAD

**Issue by Dylan-Kerler** [DLR-RM/stable-baselines3#347]
### Question

When should I use a custom feature extractor vs a custom policy? It's a little unclear here in the docs on what the differences are. If I want to use a custom neural net, should I replace the feature extractor, define it as a custom policy or, should I define both a custom policy AND a custom feature extractor?

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Feature extractors only concern themselves with processing [whatever shaped] inputs into nice 1D vectors. Policies then take this 1D vector and map it into value/pi predictions et cetera. Policy holds the feature extractor and also handles initializations and such.

So, to change most of the network, you probably want to define a new feature extractor. If your observations are 1D vectors, then you can use `net_arch` argument to change the network. If you want something more custom, however, then you need to create a custom policy.

If some part of the docs was unclear, do point it out so it can be refined for clarity.

**Comment by araffin** [comment]
I think we should update the doc. I created two diagrams to explain things faster:

![CustomPolicy](https://user-images.githubusercontent.com/1973948/110625144-5ae9ff80-819f-11eb-9ba1-761720fd6905.png)
![SB3Policy](https://user-images.githubusercontent.com/1973948/110625163-5e7d8680-819f-11eb-97a8-708050af4ef4.png)

The feature extractor is usually shared between networks to save computation (can be disabled) and the network architecture comes afterward.
As mentioned in the doc, SB3 does an abuse of language when we talk about "Policy", in the code, it refers to all the networks + optimizer + target networks, whereas in RL it refers to the actor part only (the one taking actions).

**Comment by Dylan-Kerler** [comment]
Ok I see. So if I want true customisation I should follow the "Advanced" example here: https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html#on-policy-algorithms

And then in addition to that I should *also* create a custom feature extractor.

So to summarise, if I create a feature extractor `F` and a custom policy `P` then the parameters for `F` will be shared between both the actor and the critic - that is to say, the actor and critic will use the same `F`. Then inside of `P` there are no constraints other than there must be two outputs (one for the value function and one for the policy).

Also, after your diagram I have another question. It says that the Network Architecture can be partially shared between actor/critic - What does that mean exactly? To what degree is it shared?

**Comment by Miffyli** [comment]
If you define a custom policy there is no need to do a custom feature extractor. The separation is merely done in default policies to make it clear which part processes the input into 1D vector and which part maps into pi/value, and also make it easier to change this preprocessing network without having to touch other parts.

For `P` the only real constraint is that you implement the public functions correctly (see [the original ActorCriticPolicy](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L318)). Other than that you are free to do almost anything.

**Comment by pengzhi1998** [comment]
Thank you for the clear explanations!

However, I'm still confused about a few points here. I'm using PPO actually. And from the [paper](https://arxiv.org/pdf/1707.06347.pdf), it seems that if training a network that shares parameters in actor and critic, the loss should be different:
![Screenshot from 2022-05-24 16-58-00](https://user-images.githubusercontent.com/38170947/169998508-19fdd0d4-f3ec-4bf3-a503-e4b97c42aa58.png)

Do I need to change the loss? Or, maybe an easier way is to just write a custom policy?

Thank you!
